### ISSUE TITLE
Rewrite threaded integration tests using coroutines.

### ISSUE THREAD

**Issue by Lukasa** [python-hyper/hyper-h2#10]
Like all my best ideas, this one is actually @shazow's: why use threads and sockets when we can fake the whole lot out and use coroutines for the tests in `test_interacting_stacks.py`? From IRC:

```
[19:11:34] shazow:	if i had a ~month of funded work, i'd love to rewrite urllib3's testing framework
[19:11:37] shazow:	maybe collab with you on that
[19:12:18] shazow:	ideally release it as a standalone thing that you can use to test any client/server protocol thing
[19:12:27] lukasa:	Oh that'd be super cool.
[19:12:48] lukasa:	I think I want to rewrite this without an actual socket, instead providing a 'socket-like' that is actually a queue
[19:13:01] lukasa:	Because all this messing about with events is fundamentally about ensuring that I don't leave data in sockets
[19:13:09] lukasa:	Or get stuck waiting for data that will never come
[19:13:26] shazow:	yes, exactly
[19:13:34] shazow:	a mock socket would be really powerful
[19:13:37] shazow:	and mock async-ness
[19:13:48] lukasa:	Such a neat idea
[19:13:55] lukasa:	OH CRAP
[19:13:56] shazow:	(rather than threads)
[19:13:58] lukasa:	Why am I not doing this with coroutines?
[19:14:02] lukasa:	Alright
[19:14:02] shazow:	:)
[19:14:08] lukasa:	Tearing this up and doing new stuff
```

**Comment by shazow** [comment]
My thinking was more of a ContextManager-based call queue, this way it could work on Py2 and the async-ness is nice and fake and predictable. :D

**Comment by Lukasa** [comment]
@shazow I don't mean literally with coroutines (as in yield from and asyncio and stuff), I mean *effectively* with coroutines (the plain `yield` statement). Basically, the client and server repeatedly `yield` the data they want to send, and the supervisor passes it between them.

**Comment by Lukasa** [comment]
Ok, this is fixed by 94508fe32b9279b866cbc4ca5d11dc6e7d15e9f7.

That commit is a thing of beauty, IMO. The new test harness is much smaller, just a few lines, and substantially clearer than it was before. Additionally, the tests themselves are *way* clearer to read. This was a fantastic idea, and I think we should see if we can adapt it into urllib3 (I think it's do-able, but we'll need to have a fake socket wrapper and do some trampolining to make it work. Still, should be ok.

**Comment by shazow** [comment]
That is basically exactly how I imagined it but actually functioning, beautifully done!
